[알린] Vending step6 - PR!!

jinios opened the pull request:
Step6 요구사항 구현하였습니다. 
- 자판기 메뉴의 테두리에 투명도를 줘서, 음료수의 상 하단에 비쳐 보여지는 부분이 있습니다.
- 구매 히스토리를 보여주는 기능에서, 음료객체에 따라 추가되어야 하는 UIImage를 리턴하는 동작을 BeverageImageVIew에 넣게 되었습니다. 다만 여기서 ObjectIdentifier와 switch-case문을 사용하고있는데, 이 부분이 조금 어색한 것 같습니다ㅠㅠ 힌트 부탁드려요 JK...

jinios committed:
Add Buy buttons

jinios committed:
Rename Notification Name

jinios committed:
Add fixed array in Decode, init() methods

jinios committed:
Add Notification post and selector

jinios committed:
Remove feature - updating shelf

jinios committed:
Add feature: lastPurchasedItem()
Add this feature for making purchase history view in ViewController

jinios committed:
Add lastPurchasedItem() in parent modules

jinios committed:
Rename- Notification name - updatePurchaseLog

jinios committed:
Remove Notification post

jinios committed:
Adjust position of views, re-sizing of item images

jinios committed:
Modify setBeverageImage()
Add UIImageView setting properties - border, border color, border width, adjust corner radius

jinios committed:
Add feature - Show purchase history of vending machine
- Add notification, selector
- Add purchasedItemView()

jinios committed:
Update README - step6 implements

godrm commented on VendingMachineApp/VendingMachineApp/BeverageImageView.swift:
```diff
@@ -16,9 +16,32 @@ class BeverageImageView: UIImageView {
+
+    func getItemImage(_ item: Beverage) -> BeverageImageView {
+        switch ObjectIdentifier(type(of: item)) {
+        case ObjectIdentifier(type(of: BananaMilk())):
```
ObjectIdentifier()를 안써도 클래스 타입만으로 확인할 수 있지 않나요? is 나 as 같은거요
한 번 더 생각하면 JSONParser 에서 어떻게 했었죠?

godrm commented on VendingMachineApp/VendingMachineApp/Shelf.swift:
```diff
@@ -40,13 +56,14 @@ class Shelf: NSObject, NSCoding {
         }
-        guard option > 0 else {
+        guard option > -1 else {
             throw Exception.outOfStock
```
동일한 에러 케이스라면 두 비교문을 합쳐도 되지 않나요?

godrm commented on VendingMachineApp/VendingMachineApp/BeverageImageView.swift:
```diff
@@ -16,9 +16,32 @@ class BeverageImageView: UIImageView {
+        case ObjectIdentifier(type(of: SoftDrink())):
+            return BeverageImageView(image: UIImage(named: "coke.jpg"))
+        default:
+            return BeverageImageView(image: UIImage(named: "bananamilk.jpg"))
```
여기서 반복되는 문구들을 뚫어져라 쳐다보면 매직아이처럼 겹쳐서 보이지 않을까요? 
어디까지 동일하고 어디만 다른지 생각해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Shelf.swift:
```diff
@@ -14,12 +14,28 @@ class Shelf: NSObject, NSCoding {
+            ObjectIdentifier(type(of: Coffee())),
+            ObjectIdentifier(type(of: DolceLatte())),
+            ObjectIdentifier(type(of: EnergyDrink())),
+            ObjectIdentifier(type(of: SoftDrink()))
```
여기 하드코딩들도 줄일 수 있는 방법이 있는지 고민해보세요. (정말 최선인가...)

godrm commented on VendingMachineApp/VendingMachineApp/StockController.swift:
```diff
@@ -39,22 +38,19 @@ class StockController: NSObject, NSCoding {
     func removeItem(_ itemCode: Int) throws {
         let key = try shelf.matchCode(option: itemCode)
         self.stock.removeItem(key)
-        self.shelf = shelf.update(newItems: self.stock.currentInventory())
```
위에 remove 한 뒤에 에러가 발생하는 정확한 원인을 찾았나요?
두 개로 나눈 다른 함수는 어떤건가요?

jinios replied:
buy()에 포함되어있던 동작을 removeItem()으로 따로 분리하였습니다. 

주석처럼 처리한건 레벨2에서 콘솔 자판기 만들때였고, 에러가 발생하여 동작을 분리한 이유는
StockController의 상위모듈인 VendingMachine에서 구매하는 동작은 buy() -> Beverage로 음료를 리턴하는 함수였는데,
구매된 음료를 stock에서 제거하고, "제거된 음료수를 리턴하려고하니까" out of range에러가 생기는 문제였습니다.

그래서 상위모듈인 VendingMachine의 buy()는 하위의 buy와 remove를 순서대로 호출하는 방식으로 수정해서 해결했었습니다.

godrm replied:
그렇군요. 그런 경우는 remove 할때 보통 참조하는 변수에 담아놓고 그걸 리턴해서 그대로 쓰도록 구현합니다.
다시 꺼내가기 보다는 그대로 리턴하는 방식을 생각해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -26,26 +26,36 @@ class ViewController: UIViewController {
+    @IBAction func buyButtonTouched(_ sender: UIButton) {
+        guard let item = try? vending!.buy(itemCode: sender.tag) else { return }
+//        self.updatePurchasedItemView(self.vending!.lastPurchasedItem())
+    }
```
사용안하는 코드라면 깃헙을 믿고 그냥 지우세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -74,13 +84,21 @@ class ViewController: UIViewController {
+            return
+        }
+        let yLocation = 575
+        let xLocation = 40 + (numberOfPurchase * 50)
```
하드코딩한 숫자 값을 의미있는 코드로 표현해보세요.

godrm replied:
0.73 이란 숫자가 의미하는 것을 상수로 선언해서 의도를 표현하라는 거죠.
`let xLocation = 40 + (order * 50)` 에서 에서 40은 어떤 의미고 50은 어떤 의미인지 설명하라는 겁니다.
그래야 오래 지난후에 다시 봐도 그 값이 어떤 의미인지 알수 있고, 다른 개발자가 읽을 때도 마찬가지 입니다. 
의미있는 코드라는 건 숫자를 설명할 수 있는 걸 말해요.

godrm reviewed (commented):
다음 피드백을 확인해보세요.

jinios committed:
Update README - UIView

jinios committed:
Add ImageSelector protocol

jinios committed:
Implement Extension ImageSelector

jinios committed:
Add setIntroView() - split function of initial scene

jinios committed:
Add protocol - ImageSelector
Implement ImageSelector protocol on Beverage class for sending image files' name to UIView init()

jinios committed:
Update README - UIView

jinios committed:
Add feature- Update Inventory
Resolve problem: Add feature not working- update the number of items when the stock has empty dictionary set
Fix the order of items in the shelf by modifying update() in Shelf class

jinios committed:
Delete unused code - update shelf with new dictionary set

jinios committed:
Rename- img files

jinios committed:
Delete getItemImage()

jinios committed:
Add feature - previousHistoryLog()
Execute when the view is launching (viewDidLoad())
- Check vendingMachine's purchase history is nil or empty
- If it's empty, method returns
- It the purchase history has the Beverage, add subViews with purchased item images

jinios committed:
Change encoding data
Encode empty purchase and supply log when archive data

jinios committed:
Add feature - get purchase history log

jinios committed:
Modify - Encode purchase and supply log

jinios committed:
Modify(shorten code) - setting range

jinios committed:
Modify - change subView's y value
Change y value by calculating super view's frame

jinios commented:
피드백 사항 수정했습니다. 
- ImageSelector 프로토콜 구현하여 BeverageImageView의 getImage()함수에서 switch-case제거
- 음료 클래스 이름으로 이미지를 가져올수있게 수정해서, 이미지파일의 이름을 클래스와 동일하게 수정했습니다.
- 특정 조건에서, 음료 추가나 히스토리 저장이 안됐던 문제를 수정했습니다. 
- 재고가 0개인 상태에서 음료 추가가 되면서 딕셔너리 Key가 초기화되서, Shelf를 고정된 순서의 어레이로 만들었습니다.
- Shelf를 고정 어레이로 초기화하려고 밸류들을 다 타이핑해줬는데, 순서를 고정하려면 이 방법이 제일 나은 것 같았는데 어떻게 수정해야할지 힌트를 주시면 안될까요ㅠㅠ
- 서브뷰를 추가할때 의미있는 값으로 바꾸라고하셔서 y를 변경했는데.. 의미있는 값이란 이렇게 구하는게 맞나요? x값도 감이안옵니다..!ㅠㅠ

godrm commented:
배열로 선언해서 순서를 보장하는 것 자체는 문제가 아닌 것 같네요
대신 반복되는 코드를 줄일 수 있는지 고민해보라는 거였습니다. 
- type(of: Coffee()) 는 Coffee.self 로 표현할 수 있습니다. 
`[String.self, Int.self, Double.self].forEach{ ObjectIdentifier($0) }` 이런식으로 해볼 수 있을까 했는데 [] 배열 타입이 [Any]가 되버려서 실제로는 동작 안함
각 타입에 ObjectId() 같은 메소드를 만들어서 `[ Coffee.ObjectId(), ... ]`정도는 어떨까요

jinios committed:
Add HistoryImageMaker class
- Make BeverageImageView for showing history log on the view

jinios committed:
Modify- call self.init() in decode method

jinios committed:
Modify- using map() in init()

jinios commented:
피드백 주신 사항 수정했습니다!
- Shelf 의 init()에서 map()사용으로 중복 제거
- HistoryImageMaker 객체 추가하고 구매목록 이미지 추가하는 기능 구현, 사용하는 좌표값 상수 선언

godrm commented:
수고하셨습니다. 다음 단계로 갑시다.
